### PR TITLE
feat: Add tools for open weather search and image generation using the Spark API.

### PR DESCRIPTION
**Pull request by Onelevenvy**
# Description
 Add tools for open weather search and image generation using the Spark API.

1. weather tools: using open weather api [https://openweathermap.org/api](url)

2. text 2 img tools : using the Spark API. [https://console.xfyun.cn/](url)

Fixes # (issue)

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] TODO

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods
- [ ] `optional` I have made corresponding changes to the documentation 
- [ ] `optional` I have added tests that prove my fix is effective or that my feature works
- [ ] `optional` New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by crazywoola** [comment]
Hello there, thanks for your contribution, please add my wechat: crazyphage. I will invite you to the contributors' group.

**Comment by Onelevenvy** [comment]
> Hello there, thanks for your contribution, please add my wechat: crazyphage. I will invite you to the contributors' group.
> 
> And please fix the Ci build error as well.

OK,The CI build error has been resolved.

File: api/core/tools/provider/builtin/sparkimagegeneration/tools/sparkimggen.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    edidx = host.index("/")
+    if edidx <= 0:
+        raise AssembleHeaderException("invalid request url:" + requset_url)
+    path = host[edidx:]
+    host = host[:edidx]
+    u = Url(host, path, schema)
+    return u
+
+
+# 生成鉴权url
```
Comment (Yeuoly) [inline api/core/tools/provider/builtin/sparkimagegeneration/tools/sparkimggen.py]: using full English comments is recommended which could make sense to all developers.

File: api/core/tools/provider/builtin/sparkimagegeneration/tools/sparkimggen.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+    authorization = base64.b64encode(authorization_origin.encode("utf-8")).decode(
+        encoding="utf-8"
+    )
+    values = {"host": host, "date": date, "authorization": authorization}
+
+    return requset_url + "?" + urlencode(values)
+
+
+# 生成请求body体
```
Comment (Yeuoly) [inline api/core/tools/provider/builtin/sparkimagegeneration/tools/sparkimggen.py]: The same as above

File: api/core/tools/provider/builtin/sparkimagegeneration/tools/sparkimggen.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    def __init__(self, msg):
+        self.message = msg
+
+
+class Url:
+    def __init__(this, host, path, schema):
+        this.host = host
+        this.path = path
+        this.schema = schema
+        pass
```
Comment (Yeuoly) [inline api/core/tools/provider/builtin/sparkimagegeneration/tools/sparkimggen.py]: This `pass` should be removed.

File: api/core/tools/provider/builtin/sparkimagegeneration/tools/sparkimggen.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    authorization = base64.b64encode(authorization_origin.encode("utf-8")).decode(
+        encoding="utf-8"
+    )
+    values = {"host": host, "date": date, "authorization": authorization}
+
+    return requset_url + "?" + urlencode(values)
+
+
+# 生成请求body体
+def getBody(appid, text):
```
Comment (Yeuoly) [inline api/core/tools/provider/builtin/sparkimagegeneration/tools/sparkimggen.py]: Maybe we can change this function name to `snake case`, so that all functions declared here could have the same standard.

File: api/core/tools/provider/builtin/sparkimagegeneration/tools/sparkimggen.yaml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  label:
+    en_US: Sparktex2img
+    zh_Hans: 讯飞星火图片生成
+    pt_BR: Sparktex2img
+  icon: icon.png
+  description:
+    en_US: Sparktex2img
+    zh_Hans: 讯飞星火图片生成
+    pt_BR: Sparktex2img
+
```
Comment (Yeuoly) [inline api/core/tools/provider/builtin/sparkimagegeneration/tools/sparkimggen.yaml]: There is an empty line

File: api/core/tools/provider/builtin/sparkimagegeneration/tools/sparkimggen.yaml [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    label:
+      en_US: Prompt
+      zh_Hans: 提示词
+      pt_BR: Prompt
+    human_description:
+      en_US: Image prompt
+      zh_Hans: 图像提示词
+      pt_BR: Image prompt
+    llm_description: Image prompt of Sparktex2img, you should describe the image you want to generate as a list of words as possible as detailed
+    form: llm
```
Comment (Yeuoly) [inline api/core/tools/provider/builtin/sparkimagegeneration/tools/sparkimggen.yaml]: Our ci requires a empty line at the end of a yaml file, refer to https://yamllint.readthedocs.io/en/stable/rules.html#module-yamllint.rules.new_line_at_end_of_file

File: api/core/tools/provider/builtin/weather/weather.yaml [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+      en_US: Please enter your openweathermap API Key
+      zh_Hans: 请输入你的open weather API Key
+      pt_BR: Insira sua chave de API openweathermap
+    help:
+      en_US: Get your API Key from openweathermap
+      zh_Hans: 从openweathermap获取您的 API Key
+      pt_BR: Obtenha sua chave de API do openweathermap
+    url: https://openweathermap.org
+
+
```
Comment (Yeuoly) [inline api/core/tools/provider/builtin/weather/weather.yaml]: One more line should be deleted

File: api/core/tools/provider/builtin/weather/weather.yaml [hunk 8]
```diff
@@ -0,0 +1,31 @@
+identity:
+  author: Onelevenvy
+  name: weather
```
Comment (Yeuoly) [inline api/core/tools/provider/builtin/weather/weather.yaml]: It should not be named `weather`, `open_weather` is better, because there are lots of tools have weather ability.

File: api/core/tools/provider/builtin/sparkimagegeneration/sparkimagegeneration.yaml [hunk 9]
```diff
@@ -0,0 +1,62 @@
+identity:
+  author: Onelevenvy
+  name: sparkimagegeneration
```
Comment (Yeuoly) [inline api/core/tools/provider/builtin/sparkimagegeneration/sparkimagegeneration.yaml]: `sparkimagegeneration` is confusing, I would like it to be `spark`, the followings are my reasons
1. Spark is a model provider instead of image model generation provider, there will be more models to be released which can be used as tool, for example, image understanding, ppt generation.
2. provider name should be in `snake case` standard, join all words directly just make confusions to other developers.

File: api/core/tools/provider/builtin/sparkimagegeneration/tools/sparkimggen.yaml [hunk 10]
```diff
@@ -0,0 +1,33 @@
+identity:
+  name: Sparktex2img
```
Comment (Yeuoly) [inline api/core/tools/provider/builtin/sparkimagegeneration/tools/sparkimggen.yaml]: Is it a typo? I wonder if it's `spark_text2img`

File: api/core/tools/provider/builtin/sparkimagegeneration/tools/sparkimggen.yaml [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+  author: Onelevenvy
+  label:
+    en_US: Sparktex2img
```
Comment (Yeuoly) [inline api/core/tools/provider/builtin/sparkimagegeneration/tools/sparkimggen.yaml]: The same as above

File: api/core/tools/provider/builtin/sparkimagegeneration/tools/sparkimggen.yaml [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    zh_Hans: 讯飞星火图片生成
+    pt_BR: Sparktex2img
+  
+description:
+  human:
+    en_US: Sparktex2img is a text to image tool
+    zh_Hans: 根据用户输入的文本绘制精美的图片，是文本到图片的工具，由讯飞星火提供图片生成api
+    pt_BR: Sparktex2img is a text to image tool
+  llm: parktex2img is a tool used to generate images from text
```
Comment (Yeuoly) [inline api/core/tools/provider/builtin/sparkimagegeneration/tools/sparkimggen.yaml]: Maybe a typo

**Review by Yeuoly — CHANGES_REQUESTED** [review]
After all, Thanks for your contributions! 
You can request me to review again after all changes have been done or you can add me WeChat in our contributor group, My name is `Yeuoly`.

File: api/core/tools/provider/builtin/sparkimagegeneration/tools/sparkimggen.py (hunk 4 above)
Comment (Onelevenvy) [reply api/core/tools/provider/builtin/sparkimagegeneration/tools/sparkimggen.py]: > Maybe we can change this function name to `snake case`, so that all functions declared here could have the same standard.

I resolving the issues raised in the code review. Sorry for the interruption

**Comment by Yeuoly** [comment]
hello, our designer has made a better icon for open weather, what about change to this one?
```
<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 16 16" fill="none">
  <g clip-path="url(#clip0_16624_62807)">
    <path d="M7.11111 0.888889C7.11111 0.888889 7.11111 0 8 0C8.88889 0 8.88889 0.888889 8.88889 0.888889V1.77778C8.88889 1.77778 8.88889 2.66667 8 2.66667C7.11111 2.66667 7.11111 1.77778 7.11111 1.77778V0.888889ZM15.1111 7.11111C15.1111 7.11111 16 7.11111 16 8C16 8.88889 15.1111 8.88889 15.1111 8.88889H14.2222C14.2222 8.88889 13.3333 8.88889 13.3333 8C13.3333 7.11111 14.2222 7.11111 14.2222 7.11111H15.1111ZM1.77778 7.11111C1.77778 7.11111 2.66667 7.11111 2.66667 8C2.66667 8.88889 1.77778 8.88889 1.77778 8.88889H0.888889C0.888889 8.88889 0 8.88889 0 8C0 7.11111 0.888889 7.11111 0.888889 7.11111H1.77778ZM4.05378 3.24133C4.05378 3.24133 4.68222 3.86978 4.05378 4.49822C3.42533 5.12667 2.79689 4.49822 2.79689 4.49822L2.168 3.87022C2.168 3.87022 1.53956 3.24178 2.168 2.61289C2.79689 1.98444 3.42533 2.61289 3.42533 2.61289L4.05378 3.24133ZM13.2036 4.49822C13.2036 4.49822 12.5751 5.12667 11.9467 4.49822C11.3182 3.86978 11.9467 3.24133 11.9467 3.24133L12.5751 2.61289C12.5751 2.61289 13.2036 1.98444 13.832 2.61289C14.4604 3.24133 13.832 3.86978 13.832 3.86978L13.2036 4.49822ZM3.87022 13.8316C3.87022 13.8316 3.24178 14.46 2.61333 13.8316C1.98489 13.2031 2.61333 12.5747 2.61333 12.5747L3.24178 11.9462C3.24178 11.9462 3.87022 11.3178 4.49867 11.9462C5.12711 12.5747 4.49867 13.2031 4.49867 13.2031L3.87022 13.8316Z" fill="#FFCF27"/>
    <path d="M8.00011 12.4446C10.4547 12.4446 12.4446 10.4547 12.4446 8.00011C12.4446 5.54551 10.4547 3.55566 8.00011 3.55566C5.54551 3.55566 3.55566 5.54551 3.55566 8.00011C3.55566 10.4547 5.54551 12.4446 8.00011 12.4446Z" fill="#FFCB13"/>
    <path d="M13.2343 10.3111C12.949 10.3111 12.6743 10.3556 12.4152 10.4378C12.1094 9.53647 11.2774 8.88892 10.2966 8.88892C9.24411 8.88892 8.36322 9.63469 8.11922 10.6387C7.85878 10.436 7.53744 10.3116 7.18544 10.3116C6.32633 10.3116 5.62989 11.0276 5.62989 11.9116C5.62989 12.1262 5.67255 12.3298 5.74722 12.5174C5.59878 12.4742 5.44544 12.4445 5.28411 12.4445C4.32944 12.4445 3.55566 13.2405 3.55566 14.2222C3.55566 15.204 4.32944 16 5.28411 16H13.2348C14.7619 16 16.0001 14.7271 16.0001 13.1556C16.0001 11.5845 14.7619 10.3111 13.2343 10.3111Z" fill="#E9F6FF"/>
  </g>
  <defs>
    <clipPath id="clip0_16624_62807">
      <rect width="16" height="16" fill="white"/>
    </clipPath>
  </defs>
</svg>
```

**Comment by Onelevenvy** [comment]
> hello, our designer has made a better icon for open weather, what about change to this one?
> 
> ```
> <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 16 16" fill="none">
>   <g clip-path="url(#clip0_16624_62807)">
>     <path d="M7.11111 0.888889C7.11111 0.888889 7.11111 0 8 0C8.88889 0 8.88889 0.888889 8.88889 0.888889V1.77778C8.88889 1.77778 8.88889 2.66667 8 2.66667C7.11111 2.66667 7.11111 1.77778 7.11111 1.77778V0.888889ZM15.1111 7.11111C15.1111 7.11111 16 7.11111 16 8C16 8.88889 15.1111 8.88889 15.1111 8.88889H14.2222C14.2222 8.88889 13.3333 8.88889 13.3333 8C13.3333 7.11111 14.2222 7.11111 14.2222 7.11111H15.1111ZM1.77778 7.11111C1.77778 7.11111 2.66667 7.11111 2.66667 8C2.66667 8.88889 1.77778 8.88889 1.77778 8.88889H0.888889C0.888889 8.88889 0 8.88889 0 8C0 7.11111 0.888889 7.11111 0.888889 7.11111H1.77778ZM4.05378 3.24133C4.05378 3.24133 4.68222 3.86978 4.05378 4.49822C3.42533 5.12667 2.79689 4.49822 2.79689 4.49822L2.168 3.87022C2.168 3.87022 1.53956 3.24178 2.168 2.61289C2.79689 1.98444 3.42533 2.61289 3.42533 2.61289L4.05378 3.24133ZM13.2036 4.49822C13.2036 4.49822 12.5751 5.12667 11.9467 4.49822C11.3182 3.86978 11.9467 3.24133 11.9467 3.24133L12.5751 2.61289C12.5751 2.61289 13.2036 1.98444 13.832 2.61289C14.4604 3.24133 13.832 3.86978 13.832 3.86978L13.2036 4.49822ZM3.87022 13.8316C3.87022 13.8316 3.24178 14.46 2.61333 13.8316C1.98489 13.2031 2.61333 12.5747 2.61333 12.5747L3.24178 11.9462C3.24178 11.9462 3.87022 11.3178 4.49867 11.9462C5.12711 12.5747 4.49867 13.2031 4.49867 13.2031L3.87022 13.8316Z" fill="#FFCF27"/>
>     <path d="M8.00011 12.4446C10.4547 12.4446 12.4446 10.4547 12.4446 8.00011C12.4446 5.54551 10.4547 3.55566 8.00011 3.55566C5.54551 3.55566 3.55566 5.54551 3.55566 8.00011C3.55566 10.4547 5.54551 12.4446 8.00011 12.4446Z" fill="#FFCB13"/>
>     <path d="M13.2343 10.3111C12.949 10.3111 12.6743 10.3556 12.4152 10.4378C12.1094 9.53647 11.2774 8.88892 10.2966 8.88892C9.24411 8.88892 8.36322 9.63469 8.11922 10.6387C7.85878 10.436 7.53744 10.3116 7.18544 10.3116C6.32633 10.3116 5.62989 11.0276 5.62989 11.9116C5.62989 12.1262 5.67255 12.3298 5.74722 12.5174C5.59878 12.4742 5.44544 12.4445 5.28411 12.4445C4.32944 12.4445 3.55566 13.2405 3.55566 14.2222C3.55566 15.204 4.32944 16 5.28411 16H13.2348C14.7619 16 16.0001 14.7271 16.0001 13.1556C16.0001 11.5845 14.7619 10.3111 13.2343 10.3111Z" fill="#E9F6FF"/>
>   </g>
>   <defs>
>     <clipPath id="clip0_16624_62807">
>       <rect width="16" height="16" fill="white"/>
>     </clipPath>
>   </defs>
> </svg>
> ```
> 
> and here is what it looks like image


Hello, Ok，I will update it later. Thank you!

**Comment by Yeuoly** [comment]
> hello, our designer has made a better icon for open weather, what about change to this one?
> 
> ```
> <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 16 16" fill="none">
>   <g clip-path="url(#clip0_16624_62807)">
>     <path d="M7.11111 0.888889C7.11111 0.888889 7.11111 0 8 0C8.88889 0 8.88889 0.888889 8.88889 0.888889V1.77778C8.88889 1.77778 8.88889 2.66667 8 2.66667C7.11111 2.66667 7.11111 1.77778 7.11111 1.77778V0.888889ZM15.1111 7.11111C15.1111 7.11111 16 7.11111 16 8C16 8.88889 15.1111 8.88889 15.1111 8.88889H14.2222C14.2222 8.88889 13.3333 8.88889 13.3333 8C13.3333 7.11111 14.2222 7.11111 14.2222 7.11111H15.1111ZM1.77778 7.11111C1.77778 7.11111 2.66667 7.11111 2.66667 8C2.66667 8.88889 1.77778 8.88889 1.77778 8.88889H0.888889C0.888889 8.88889 0 8.88889 0 8C0 7.11111 0.888889 7.11111 0.888889 7.11111H1.77778ZM4.05378 3.24133C4.05378 3.24133 4.68222 3.86978 4.05378 4.49822C3.42533 5.12667 2.79689 4.49822 2.79689 4.49822L2.168 3.87022C2.168 3.87022 1.53956 3.24178 2.168 2.61289C2.79689 1.98444 3.42533 2.61289 3.42533 2.61289L4.05378 3.24133ZM13.2036 4.49822C13.2036 4.49822 12.5751 5.12667 11.9467 4.49822C11.3182 3.86978 11.9467 3.24133 11.9467 3.24133L12.5751 2.61289C12.5751 2.61289 13.2036 1.98444 13.832 2.61289C14.4604 3.24133 13.832 3.86978 13.832 3.86978L13.2036 4.49822ZM3.87022 13.8316C3.87022 13.8316 3.24178 14.46 2.61333 13.8316C1.98489 13.2031 2.61333 12.5747 2.61333 12.5747L3.24178 11.9462C3.24178 11.9462 3.87022 11.3178 4.49867 11.9462C5.12711 12.5747 4.49867 13.2031 4.49867 13.2031L3.87022 13.8316Z" fill="#FFCF27"/>
>     <path d="M8.00011 12.4446C10.4547 12.4446 12.4446 10.4547 12.4446 8.00011C12.4446 5.54551 10.4547 3.55566 8.00011 3.55566C5.54551 3.55566 3.55566 5.54551 3.55566 8.00011C3.55566 10.4547 5.54551 12.4446 8.00011 12.4446Z" fill="#FFCB13"/>
>     <path d="M13.2343 10.3111C12.949 10.3111 12.6743 10.3556 12.4152 10.4378C12.1094 9.53647 11.2774 8.88892 10.2966 8.88892C9.24411 8.88892 8.36322 9.63469 8.11922 10.6387C7.85878 10.436 7.53744 10.3116 7.18544 10.3116C6.32633 10.3116 5.62989 11.0276 5.62989 11.9116C5.62989 12.1262 5.67255 12.3298 5.74722 12.5174C5.59878 12.4742 5.44544 12.4445 5.28411 12.4445C4.32944 12.4445 3.55566 13.2405 3.55566 14.2222C3.55566 15.204 4.32944 16 5.28411 16H13.2348C14.7619 16 16.0001 14.7271 16.0001 13.1556C16.0001 11.5845 14.7619 10.3111 13.2343 10.3111Z" fill="#E9F6FF"/>
>   </g>
>   <defs>
>     <clipPath id="clip0_16624_62807">
>       <rect width="16" height="16" fill="white"/>
>     </clipPath>
>   </defs>
> </svg>
> ```
> 
> and here is what it looks like image

you can also find a SVG formatted icon in `core.model_runtime.model_providers.spark._assets.icon_s_en.svg`, it brings smaller size and better clarity.

File: sdks/python-client/dify_client/__init__.py [hunk 13]
```diff
@@ -1 +0,0 @@
-from dify_client.client import ChatClient, CompletionClient, DifyClient
```
Comment (Yeuoly) [inline sdks/python-client/dify_client/__init__.py:1]: what is this modification made for?

File: api/core/model_runtime/model_providers/tongyi/_common.py [hunk 14]
```diff
@@ -1,4 +1,11 @@
-from core.model_runtime.errors.invoke import InvokeError
```
Comment (Yeuoly) [inline api/core/model_runtime/model_providers/tongyi/_common.py]: these modifications look like not related to this PR, you can create another one instead.

**Review by Yeuoly — CHANGES_REQUESTED** [review]

File: sdks/python-client/dify_client/__init__.py (hunk 13 above)
Comment (Onelevenvy) [reply sdks/python-client/dify_client/__init__.py:1]: I accidentally used the "ruff check . --fix" command in the API path, and it automatically corrected the imports errors.

**Review by crazywoola — APPROVED** [review]

**Review by Yeuoly — APPROVED** [review]
LGTM
